### PR TITLE
[Feature] Spell Ranks will now work with AllowSpellMemorizeFromItem Rule

### PR DESCRIPTION
**Pull request by Aeadoin**
Functions same as scribing to spellbook, 

Will check if it's a valid upgrade (no downgrading possible) and present the player with a Yes/No Popup Prompt,

unknown

### CODE REVIEW

File: zone/client_packet.cpp [hunk 1]
```diff
@@ -8916,12 +8916,30 @@ void Client::Handle_OP_ItemVerifyRequest(const EQApplicationPacket *app)
...
+				spell_id = item->Scroll.Effect;
+				if (RuleB(Spells, AllowSpellMemorizeFromItem)) 
+				{
+					int highest_spell_id = GetHighestScribedSpellinSpellGroup(spells[spell_id].spell_group);
+					if (spells[spell_id].spell_group > 0 && highest_spell_id > 0) 
+					{
+						if (spells[spell_id].rank > spells[highest_spell_id].rank)
+						{
+							std::string message = fmt::format("{} will replace {} in your spellbook", GetSpellName(spell_id), GetSpellName(highest_spell_id));
+							this->SetEntityVariable("slot_id",itoa(slot_id));
```
Comment (Akkadius) [inline zone/client_packet.cpp:8928]: Remove `this->` as its implied

Comment (Akkadius) [reply zone/client_packet.cpp:8928]: For the below lines as well

File: zone/client_packet.cpp [hunk 2]
```diff
@@ -8916,12 +8916,30 @@ void Client::Handle_OP_ItemVerifyRequest(const EQApplicationPacket *app)
...
+							this->SetEntityVariable("spell_id",itoa(item->ID));
+							this->SendPopupToClient("", message.c_str(), 1000001, 1, 10);
+							return;
+						} 
+						else if (spells[spell_id].rank < spells[highest_spell_id].rank)
+						{
+							MessageString(Chat::Red, 11004, spells[spell_id].name, spells[highest_spell_id].name);
```
Comment (Akkadius) [inline zone/client_packet.cpp:8935]: DB String ID 11004 needs to be referenced as a constant

File: zone/client.h [hunk 3]
```diff
@@ -208,6 +208,7 @@ const std::string DIAWIND_RESPONSE_TWO_KEY       = "diawind_npc_response_two";
 const uint32      POPUPID_DIAWIND_ONE            = 99999;
 const uint32      POPUPID_DIAWIND_TWO            = 100000;
 const uint32      POPUPID_UPDATE_SHOWSTATSWINDOW = 1000000;
+const uint32      POPUPID_REPLACE_SPELLWINDOW    = 1000001;
```
Comment (Akkadius) [inline zone/client.h:211]: Nice

File: zone/client_packet.cpp [hunk 4]
```diff
@@ -8916,12 +8916,30 @@ void Client::Handle_OP_ItemVerifyRequest(const EQApplicationPacket *app)
...
+							return;
+						} 
+					}
 					DeleteItemInInventory(slot_id, 1, true);
 					MemorizeSpellFromItem(item->ID);
-				} else {
```
Comment (Akkadius) [inline zone/client_packet.cpp:8922]: Removing the else intentional here ?

**Review by Akkadius — CHANGES_REQUESTED** [review]
Neat feature. Pending comments

File: zone/client_packet.cpp (hunk 4 above)
Comment (Aeadoin) [reply zone/client_packet.cpp:8922]: Yes, this is intentional.

File: zone/client.h (hunk 3 above)
Comment (Aeadoin) [reply zone/client.h:211]: Thanks!

File: zone/client_packet.cpp [hunk 5]
```diff
@@ -11122,11 +11140,20 @@ void Client::Handle_OP_PopupResponse(const EQApplicationPacket *app)
...
 
+	//Get Item Details if POPUPID_REPLACE_SPELLWINDOW was used
+
 	/**
 	 * Handle any EQEmu defined popup Ids first
 	 */
 	std::string response;
 	switch (popup_response->popupid) {
+		case POPUPID_REPLACE_SPELLWINDOW:
+				LogDebug("Are we even receiving the reponse, with item? slot [{}] spellID [{}]",std::stoi(this->GetEntityVariable("slot_id")),std::stoi(this->GetEntityVariable("spell_id")));
```
[the remark this answers is not in the document]
Comment (Aeadoin) [reply zone/client_packet.cpp:11151]: Removed (was actually a leftover log message from my testing)

File: zone/client_packet.cpp (hunk 1 above)
Comment (Aeadoin) [reply zone/client_packet.cpp]: Fixed.

File: zone/client_packet.cpp (hunk 2 above)
Comment (Aeadoin) [reply zone/client_packet.cpp]: Done, thanks!

File: zone/client_packet.cpp (hunk 4 above)
Comment (Aeadoin) [reply zone/client_packet.cpp:8922]: Added back in, since it logically makes sense to return if the rule isn't in use.

**Review by Akkadius — APPROVED** [review]

**Pull request merged**